assignments/HW4.md

```diff
@@ -0,0 +1,194 @@
+
+The existing model in the repository is very simple. Once you have familiarized yourself with the dataset, explore this model using whatever means you think are appropriate and determine in which ways you don’t believe it to be suitable. Based on what you learn in this exploratory step, you should train a new model and build a deployable microservice.
+
+You should evaluate your new model, and should explain how this model (overall) performs better than the baseline model, and any ways in which your model performs worse than the baseline model.
+
+Plan some testing, perform that testing on your microservice endpoints, and document it. At least some of this testing should be automated using Github Actions.
+As you have in previous assignments, you can create a test suite (ie. [pytest](https://flask.palletsprojects.com/en/2.2.x/testing/)) and have your workflow run the test suite.
+
+The deliverable for this task is a documented technical artifact; submit a link of your github repository to the Gradescope assignment. This should include:
+
+1. a trained model that is deployable as a microservice; the microservice must take as input a query request with student data and return a prediction about that student’s graduate school success. This is the API for your microservice. It should also include
+
+2. an updated README that clearly and concisely explains
+
+   - Clear documentation for your API. (How it should be called, what data it expects, any pre-conditions for the service, and how to understand the output).
+
+   - A description of which features you used in training your model, and how your retrained the model performs better than the baseline model
+
+   - testing and deployment instructions; and
+
+   - an explanation and justification of the testing you have done on it.
+
+### (3) Design document - (due Thursday, Nov 3rd, 23:59) | 70 points
+
+Once you have implemented the feature in isolation, your next (group) task is to document it and argue for how it should be integrated into the design of the existing system. Effectively, your technical contact’s demand for microservices makes it straightforward to develop this feature in isolation, but it raises questions about how it should be integrated into your existing (relatively monolithic) web app. You have at least several options:
+
+1.  Rewrite the web app using a different architecture (re-architect the overall app in some other way)
+
+2.  Rewrite the feature to fit more naturally into the Java/Maven-imposed architecture.
+
+3.  Mash a microservice into the Java/Maven architecture using some other mechanism.
```

Comment:
The language of the existing web app shouldn't really matter. Since we've given them starter code and asked them to implement an ML service, the only sensible answer is to just call out to that service (e.g., by writing a lightweight Java API wrapper/proxy around the service).